File: src/my_cpp_pkg/src/hw_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/hardware_status.hpp"

class HardwareStatusPublisher : public rclcpp::Node
{
public:
    HardwareStatusPublisher() : Node("hardware_status_publisher")
    {
        hw_status_publisher_ = this->create_publisher<my_robot_interfaces::msg::HardwareStatus>("hardware_status", 10);
        timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                         std::bind(&HardwareStatusPublisher::publishNumber, this));
        RCLCPP_INFO(this->get_logger(), "Number Publisher has been started");
    }

private:
    void publishNumber()
    {
        auto msg = my_robot_interfaces::msg::HardwareStatus();
        msg.temperature = 32;
        msg.are_motors_ready = true;
        msg.debug_message = "Nothing special here";
        hw_status_publisher_->publish(msg);
    }

    rclcpp::Publisher<my_robot_interfaces::msg::HardwareStatus>::SharedPtr hw_status_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    //shared pointer, you don't need to delete manually the instance when object is not referenced (like GC in Java)
    auto node = std::make_shared<HardwareStatusPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/hw_status_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from my_robot_interfaces.msg import HardwareStatus

class HardwareStatusPublisherNode(Node):

    def __init__(self):
        super().__init__("hardware_status_publisher") 
        self.hw_status_publisher_ = self.create_publisher(HardwareStatus, "hardware_status", 10) # create topic robot_news in which robot_news will publish
        self.timer = self.create_timer(1.0, self.publish_hw_status) # publish 2 times per sec 
        self.get_logger().info("Hardware Status Publisher has been started")
    
    def publish_hw_status(self):
        msg = HardwareStatus()
        msg.temperature = 32
        msg.are_motors_ready = True
        msg.debug_message = "Nothing special here"
        self.hw_status_publisher_.publish(msg)
    

def main(args=None):
    rclpy.init(args=args) # first line to be written in any ROS2 .py node
    node = HardwareStatusPublisherNode()
    rclpy.spin(node) # allow node to continue to be alive
    rclpy.shutdown() # last line of any ROS2 .py node

if __name__ == "__main__":
    main()